pkg/util/tracing/tracer_test.go

```diff
@@ -388,3 +389,27 @@ func TestTracer_VisitSpans(t *testing.T) {
 	require.Len(t, tr1.activeSpans.m, 0)
 	require.Len(t, tr2.activeSpans.m, 0)
 }
+
+// TestActiveSpanVisitorErrors confirms that the visitor of the Tracer's
+// activeSpans registry gracefully exits upon receiving a sentinel error from
+// `iterutil.StopIteration()`.
+func TestActiveSpanVisitorErrors(t *testing.T) {
+	tr := NewTracer()
+	root := tr.StartSpan("root", WithForceRealSpan())
+	child := tr.StartSpan("root.child", WithParentAndAutoCollection(root))
+	remoteChild := tr.StartSpan("root.remotechild", WithParentAndManualCollection(child.Meta()))
+
+	var numVisited int
+
+	visitor := func(*Span) error {
+		numVisited++
+		return iterutil.StopIteration()
+	}
+
+	require.NoError(t, tr.VisitSpans(visitor))
+	require.Equal(t, 1, numVisited)
+
+	remoteChild.Finish()
```

Comment:
We should simply `defer <span>.Finish()` right after starting it, that's the pattern we'd typically expect and this way we'd clean up after ourselves even if the test finished.

Reply:
It looks a lot cleaner this way, thank you!